controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1173,6 +1180,65 @@ private long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegments
         });
     }
 
+    @Override
+    public CompletableFuture<Void> recordCommitOffsets(final UUID txnId, final Map<Long, Long> commitOffsets) {
+        // The transaction may already have been committed and its record removed and this could be an idempotent run. 
```

Comment:
How does this handle the case were the transaction was committed and its record was removed but this was not successfully run?

Reply:
Txn record removal is the last step. Record commit offsets is an intermediate step during commit workflow. So if this was not run successfully then txn removal would not have happened. 

Here is the flow:
we first merge transaction segments. Then we fetch the size. Then update the commitOffsets in transaction's metadata. 
Once we have completed above operations for a batch, for each transaction in the committing set, we generate their marks and then move them to committed state (remove activeTxnRecord and create a new CompletedTxnRecord).